src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -413,65 +312,167 @@ public ValueTask<(Solution solution, bool updated)> TryUpdateWorkspaceCurrentSol
-                _cancellationTokenSource.Dispose();
+                private readonly ChecksumToSolutionCache _cache;
+                private readonly Checksum _solutionChecksum;
+                private readonly CancellationTokenSource _cancellationTokenSource = new();
+                private readonly Task<Solution> _task;
+
+                private int _refCount = 1;
+
+                // For assertion purposes.
+                public int RefCount => _refCount;
+
+                public RefCountedSolution(
+                    ChecksumToSolutionCache cache,
+                    Checksum solutionChecksum,
+                    Func<CancellationToken, Task<Solution>> getSolutionAsync)
+                {
+                    _cache = cache;
+                    _solutionChecksum = solutionChecksum;
+                    _task = getSolutionAsync(_cancellationTokenSource.Token);
+                }
+
+                public Task<Solution> Task => _task;
+
+                public void AddReference_TakeLock()
+                {
+                    using (_cache._gate.DisposableWait(CancellationToken.None))
+                    {
+                        AddReference_WhileAlreadyHoldingLock();
+                    }
+                }
 
-                await _cleanupAsync().ConfigureAwait(false);
+                public void AddReference_WhileAlreadyHoldingLock()
+                {
+                    Contract.ThrowIfFalse(_cache._gate.CurrentCount == 0);
+                    Contract.ThrowIfTrue(_refCount <= 0);
+                    _refCount++;
+                }
 
-                // Make sure to wait for the underlying asynchronous operation to complete before returning. Use a
-                // no-throw awaitable to avoid throwing an exception on cancellation or error (the inner operation is
-                // responsible for its own error reporting, if any).
-                await _task.NoThrowAwaitable(false);
+                public void Dispose()
+                {
+                    using (_cache._gate.DisposableWait(CancellationToken.None))
+                    {
+                        Release_WhileAlreadyHoldingLock();
+                    }
+                }
+
+                public void Release_WhileAlreadyHoldingLock()
+                {
+                    Contract.ThrowIfFalse(_cache._gate.CurrentCount == 0);
+                    Contract.ThrowIfTrue(_refCount <= 0);
+                    _refCount--;
+                    if (_refCount == 0)
+                    {
+                        _cancellationTokenSource.Cancel();
+                        _cancellationTokenSource.Dispose();
+
+                        Contract.ThrowIfFalse(_cache._solutionChecksumToLazySolution.TryGetValue(_solutionChecksum, out var existingSolution));
+                        Contract.ThrowIfFalse(existingSolution == this);
+                        _cache._solutionChecksumToLazySolution.Remove(_solutionChecksum);
+                    }
```

Comment:
much simpler to just keep refcount management and cache-dictionary management tied together.